### ISSUE TITLE
identityHashCode creates an `$identityHash` property on JS objects in dart2js

### ISSUE THREAD

**Issue by greglittlefield-wf** [dart-lang/sdk#47595]
Note: This is closely related to https://github.com/dart-lang/sdk/issues/36354 and I wasn't sure if I should just comment there, but the end issue I'm experiencing is different, so I figured a separate issue made more sense.

In dart2js, passing a JS object into `identityHashCode` results in an enumerable `$identityHash` property being added to it, which shows up when JS code consuming that object uses `Object.keys` or JS for-in loops, which is undesireable.

DDC, on the other hand, stores the hashcode [under a JS Symbol](https://github.com/dart-lang/sdk/blob/c74dd07a4a3bc139c6a3b6e21eee79ac4a6440dd/sdk/lib/_internal/js_dev_runtime/patch/core_patch.dart#L40), which, while still enumerable, doesn't show up when using `Object.keys` or JS for-in loops (and only shows up in `Object.getOwnPropertySymbols` or `Object.getOwnPropertyDescriptors`).

Reduced test case, reproducible on Dart 2.14.4:
```dart
import 'dart:html';
import 'dart:js_util';

void main() {
  final jsObject = newObject();
  identityHashCode(jsObject);
  window.console.dir(jsObject);
}
```
Inspecting the logged object to see the property:
<img width="305" alt="Screen Shot 2021-11-01 at 2 35 45 PM" src="https://user-images.githubusercontent.com/1713967/139745314-972ce5f6-5609-444d-992d-e614b5d52e33.png">

Ideally, `identityHashCode` would use `WeakMap` (as mentioned in https://github.com/dart-lang/sdk/issues/36354#issuecomment-477304153) so that the JS object doesn't get modified at all and this property is inaccessible from consuming JS code. If that's too much effort to change, then perhaps the dart2js implementation could be updated to use a JS Symbol so that it's consistent with the DDC implementation?

For now, we can work around the issue in most cases by avoiding the use of `identityHashCode`, by using an implementation of `jsify` that doesn't use `Map.identity()`, but it would nice to know that this property will never get added unexpectedly if code outside of our control uses `identityHashCode` with that object.

**Comment by fishythefish** [comment]
I think the main concern with `WeakMap` is that it may introduce unacceptable overhead. We landed a [change](https://github.com/dart-lang/sdk/commit/8ae984c54ab36a05422af6f250dbaa7da70fc461) to make the `$ti` property a Symbol rather than enumerable (cf. #40535), so a similar fix may be appropriate here.

/cc @rakudrama

**Comment by rakudrama** [comment]
Fixed to use symbol in 8bbf5c9586316684b5f2bbf7222a845151a47ce5.

**Comment by greglittlefield-wf** [comment]
Thank you! ❤️
